storage/rocksdb/ha_rocksdb.cc

```diff
@@ -9338,140 +9253,297 @@ int ha_rocksdb::index_read_map(uchar *const buf, const uchar *const key,
+    start_has_inf_suffix= false;
+    end_has_inf_suffix= true;
+  }
+  else if (find_flag == HA_READ_BEFORE_KEY) {
+    /*
+      We get here for queries like
+        select * from t1
+        where              pk <1007 order by pk desc limit 2 for update
+        select * from t1
+        where pk >=800 and pk <1007 order by pk desc limit 2 for update
+    */
+
+    // Move the right endpoint from slice to end_slice
+    end_slice= slice;
+
+    if (end_key) {
+      uchar pack_buffer[MAX_KEY_LENGTH];
+      uint end_slice_size=
+          kd.pack_index_tuple(table, pack_buffer, end_slice_buf,
+                              end_key->key, end_key->keypart_map);
+
+      slice= rocksdb::Slice(reinterpret_cast<char *>(end_slice_buf),
+                            end_slice_size);
+
+      end_has_inf_suffix= false;
+      big_range= false;
+    } else {
+      uint end_slice_size;
+      kd.get_infimum_key(end_slice_buf, &end_slice_size);
+      slice= rocksdb::Slice((char*)end_slice_buf, end_slice_size);
+
+     big_range= true;
     }
   }
-  return HA_EXIT_SUCCESS;
+  else if (end_key) {
+    // Known start range bounds: HA_READ_KEY_OR_NEXT, HA_READ_AFTER_KEY
+    if (find_flag == HA_READ_KEY_OR_NEXT)
+      start_has_inf_suffix= false;
+    else if (find_flag == HA_READ_AFTER_KEY)
+      start_has_inf_suffix= true;
+    else
+      DBUG_ASSERT(0);
+
+    // Known end range bounds: HA_READ_AFTER_KEY, HA_READ_BEFORE_KEY
+    if (end_key->flag == HA_READ_AFTER_KEY) {
+      // this is "key_part <= const".
+      end_has_inf_suffix= true;
+    } else if (end_key->flag == HA_READ_BEFORE_KEY) {
+      // this is "key_part < const", non-inclusive.
+      end_has_inf_suffix= false;
+    } else
+      DBUG_ASSERT(0);
+
+    uchar pack_buffer[MAX_KEY_LENGTH];
+    uint end_slice_size= kd.pack_index_tuple(table, pack_buffer, end_slice_buf,
+                                             end_key->key,
+                                             end_key->keypart_map);
+
+    end_slice= rocksdb::Slice(reinterpret_cast<char *>(end_slice_buf),
+                              end_slice_size);
+  }
+  else
+  {
+    big_range= true;
+#if 0
+    // The below is code to handle this without LockingIterator:
+    // No end key
+    // Known start range bounds: HA_READ_KEY_OR_NEXT, HA_READ_AFTER_KEY
+    if (find_flag == HA_READ_KEY_OR_NEXT)
+      start_has_inf_suffix= false;
+    else if (find_flag == HA_READ_AFTER_KEY)
+      start_has_inf_suffix= true;
+    else
+      DBUG_ASSERT(0);
+
+    uint end_slice_size;
+    kd.get_infimum_key(end_slice_buf, &end_slice_size);
+    end_slice= rocksdb::Slice((char*)end_slice_buf, end_slice_size);
+    end_has_inf_suffix= true;
+#endif
+  }
+
+  if (big_range)
+  {
+    *use_locking_iterator= true;
+    return 0;
+  }
+
+  rocksdb::Endpoint start_endp;
+  rocksdb::Endpoint end_endp;
+
+  if (kd.m_is_reverse_cf) {
+    // Flip the endpoints
+    start_endp =rocksdb::Endpoint(end_slice, !end_has_inf_suffix);
```

Comment:
It's not clear why the suffixes are inverted here?

If the forward cf range to be locked is (pk=10, 0) -> (pk=12, +inf), then isn't the reverse cf range to be locked
(pk=12, +inf) -> (pk=10, 0)? It seems (pk=12, 0) -> (pk=10, +inf) would not be right?

Reply:
The reason suffixes are inverted is as follows:

Let's consider a forward-ordered CF. The endpoint values would compare as follows (here LOWEST=0, HIGHEST=1, missing inf_suffix means any inf_suffix value):
 
```
 key=a, inf_suffix=LOWEST
 key=ab
 key=az
 key=a, inf_suffix=HIGHEST

 key=b, inf_suffix=LOWEST
 key=ba
 key=bz
 key=b, inf_suffix=HIGHEST
 ...
```
Putting the same set of keys into a reverse CF:

```
 key=b, inf_suffix=HIGHEST
 key=bz
 key=ba
 key=b, inf_suffix=LOWEST

 key=a, inf_suffix=HIGHEST
 key=az
 key=ab
 key=a, inf_suffix=LOWEST.
```
Note that the flag values have swapped their meanings, HIGHEST comes before LOWEST.

However, the endpoint comparison function in RocksDB doesn't "know" if the CF is reverse-ordered or not (it just makes use of the Key Comparator).

As for inf_suffix value, `Endpoint(key=$VAL, inf_suffix=0)` always compares before any endpoint with key=$VAL, and `Endpoint(key=$VAL, inf_suffix=1)` always compares after any endpoint with key=$VAL.

This is why one needs to flip the inf_suffix values.

Reply:
... added a comment about this.